haystack_experimental/components/preprocessors/md_header_level_inferrer.py

```diff
@@ -0,0 +1,110 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import re
+
+from haystack import Document, component, logging
+
+
+@component
+class MarkdownHeaderLevelInferrer:
+    """
+    Infers and rewrites header levels in markdown text to normalize hierarchy.
+    """
+
+    def __init__(self):
+        """
+        Initializes the MarkdownHeaderLevelInferrer.
+
+        Uses a hardcoded regex pattern to match markdown headers from level 1 to 6.
+        """
+        self._header_pattern = r"(?m)^(#{1,6}) (.+)$"
+
+    @component.output_types(documents=list[Document])
+    def run(self, documents: list[Document]) -> dict:
+        """
+        Infers and rewrites the header levels in the content for documents that use uniform header levels.
+
+        Args:
+            documents (list[Document]): List of Document objects to process.
+
+        Returns:
+            dict: A dictionary with the key 'documents' containing the processed Document objects.
+        """
+        logger = logging.getLogger(__name__)
```

Comment:
logger is usually global, you can move this to top of the file, and keep it below the import statments